frontend/src/tests/application-tests.js

```diff
@@ -104,32 +217,162 @@ export function applicationsTests({ apiGET, apiPOST }) {
-        );
-        it.todo(
-            "When submitting survey.js data an applicant and application are updated if they already exist"
-        );
-        it.todo(
-            "Even if a different utorid is submitted via survey.js data the active_user's utorid is used"
+        it.skip("Survey.js posting data is pre-filled based on prior application", async () => {
+            // Submit an application to the posting
+            const application = {
+                ...applicant,
+                program: "P",
+                program_start: "2021-06-21",
+                department: "cs",
+                previous_university_ta: false,
+                previous_department_ta: false,
+                previous_other_university_ta: false,
+                comments: "My comments",
+            };
+
+            let resp = await apiPOST(
+                `/public/postings/${postingData.url_token}/submit`,
+                application,
+                true
+            );
+            expect(resp).toHaveStatus("success");
+            const submittedData = resp.payload;
+
+            const applicationPrefillData = {
+                program: submittedData.program,
+                department: submittedData.department,
+                yip: submittedData.yip,
+                previous_department_ta: submittedData.previous_department_ta,
+                previous_university_ta: submittedData.previous_university_ta,
+                previous_experience_summary:
+                    submittedData.previous_experience_summary,
+                gpa: submittedData.gpa,
+                comments: submittedData.comments,
+            };
+
+            await switchToUser(userCreatedFromApplicant);
+
+            // Read survey.js posting data
+            resp = await apiGET(
+                `/public/postings/${postingData.url_token}`,
+                true
+            );
+            expect(resp).toHaveStatus("success");
+            checkPropTypes(surveyPropTypes, resp.payload);
+            expect(resp.payload["prefilled_data"]).toEqual(
+                expect.objectContaining(applicationPrefillData)
+            );
+            await restoreDefaultUser();
+        });
+
+        it(
+            "Submitting an application for a posting that does not have any positions or position preferences" +
+                " responds in error",
+            async () => {
+                // Create a new posting
+                const posting = {
+                    name: "CSC343F TA",
+                    intro_text: "Testing posting for CSC343F",
+                    open_date: "2021/04/01",
+                    close_date: "2021/05/01",
+                    availability: "open",
+                };
+                const respPostingData = await addPosting(posting, session);
+                Object.assign(posting, respPostingData);
+
+                // Submit an application to a posting that has no positions
+                // See issue #630
```

Comment:
The issue isn't submitting to an application with no positions. It is submitting a position preference for a position_code that isn't associated with the current application.

Reply:
I have updated both the issue and the comment itself.